src/abstractdataframe/subset.jl

```diff
@@ -125,7 +136,17 @@ function _get_subset_conditions(df::Union{AbstractDataFrame, GroupedDataFrame},
         end
     end
 
-    @assert eltype(cond) === Bool
+    # we special case 0-length cond, as in this case broadcasting does not
+    # guarantee setting a proper eltype for the result
+    if isempty(cond)
+        if eltype(cond) !== Bool
+            throw(ArgumentError("passed conditions produce $(eltype(cond)) " *
+                                "as element type of the result while only " *
+                                "Bool is allowed."))
+        end
+    else
+        @assert eltype(cond) === Bool
+    end
```

Comment:
I don't understand this. Contrary to what the comment says, the check is dependent on the eltype even for empty vectors.

Reply:
When `cond` is non-empty then broadcasting mechanism guarantees that in line 148 must hold. That is why we have `@assert` there.

Conversely, if `cond` is empty broadcasting lets incorrect element type slip through:
```
julia> subset(DataFrame(), [] => ByRow(() -> "aaa"))
ERROR: ArgumentError: passed conditions produce Union{} as element type of the result while only Bool is allowed.
```
As you can see here although the condition returns `String` it is not correctly identified as incorrect by `_and.(cols...)` call because the output vector is empty.

Is it clear now?

Reply:
But won't that error also be thrown when `cond` is empty and `broadcast` uses eltype `Any` because inference fails?

Reply:
I do not think it is possible. Can you think of an example? Here are things that I have checked:
```
julia> DataFrames._and.([]) # supertype
0-element BitVector

julia> DataFrames._and.(Number[]) # supertype
0-element BitVector

julia> DataFrames._and.(Union{Bool, Missing}[]) # supertype
0-element BitVector

julia> DataFrames._and.(Int[]) # empty intersection
Union{}[]

julia> DataFrames._and.(Char[]) # empty intersection
Union{}[]

julia> DataFrames._and.(Union{Int, Missing}[]) # empty intersection
Union{}[]
```
and my rule is the best we can count on AFAICT.

Reply:
I don't know how it's possible that inferences guesses that the result is `Bool` in so difficult cases. In theory there's no guaranty that the return type is inferred, but since I can't come up with an example...

Reply:
The empty container is not likely in practice, so I prefer to be on the safe side.